Fix torch.full with dynamic tensor fill_value in torch.compile

amaldevh opened the pull request:
Fixes #166253

## Summary
When `torch.full` is called with a 0-D tensor as `fill_value` inside a `torch.compile`'d function, the value was being incorrectly cached, causing subsequent calls with different values to return the first value.

## Root Cause
The Dynamo handler for `torch.full` was calling `aten._local_scalar_dense` to convert tensor fill_values to Python scalars at compile time, which baked the value into the compiled graph as a constant.

## Solution
Modified the Dynamo handler to decompose `torch.full(size, tensor_fill_value)` into `empty(size).fill_(tensor_fill_value)` when `fill_value` is a `TensorVariable`, keeping the fill value dynamic in the compiled graph.

## Testing
Added test case that verifies torch.full works correctly with dynamic tensor fill_values across multiple calls and dtypes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

amaldevh committed:
Fix torch.full with dynamic tensor fill_value in torch.compile

Fixes #166253

When torch.full is called with a 0-D tensor as fill_value inside a
torch.compile'd function, the value was being incorrectly converted
to a Python scalar at compile time using _local_scalar_dense. This
caused the scalar value to be baked into the compiled graph, making
subsequent calls with different fill_values incorrectly reuse the
first value.

The fix decomposes torch.full(size, tensor_fill_value) into
empty(size).fill_(tensor_fill_value) when fill_value is a TensorVariable,
keeping the fill value dynamic in the compiled graph.

Test Plan: Added test_full_with_tensor_fill_value

AmalDevHaridevan commented:
@pytorchbot label "topic: not user facing"

Lucaskabela reviewed (changes requested):
Thanks for the change! One minor nit on test coverage to make sure this implementation works in the pre-existing path correctly, but other than that it should be good so feel free to @ me when ready for another review

amaldevh committed:
Fix torch.full with dynamic tensor fill_value in torch.compile

Fixes #166253

When torch.full is called with a 0-D tensor as fill_value inside a
torch.compile'd function, the value was being incorrectly converted
to a Python scalar at compile time using _local_scalar_dense. This
caused the scalar value to be baked into the compiled graph, making
subsequent calls with different fill_values incorrectly reuse the
first value.

The fix decomposes torch.full(size, tensor_fill_value) into
empty(size).fill_(tensor_fill_value) when fill_value is a TensorVariable,
keeping the fill value dynamic in the compiled graph. Scalar fill_values
continue to use the original code path via aten.full.default.

Test Plan:
- Added test_full_with_tensor_fill_value covering:
  * Dynamic tensor fill_values across multiple calls
  * Multiple dtypes with tensor fill_values
  * Scalar fill_values (existing behavior)
  * Multiple scalar values in same function

AmalDevHaridevan commented on test/dynamo/test_functions.py:
```diff
@@ -5215,6 +5215,37 @@ def forward(self, x):
+            func_typed_compiled = torch.compile(func_typed)
+            x_typed = torch.tensor(7, dtype=dtype)
+            result = func_typed_compiled(x_typed)
+            expected = torch.full((3,), x_typed, dtype=dtype)
```
@Lucaskabela  Updated with test coverage for non-tensor fill_values! 

The tests now cover:
1. ✅ Dynamic tensor fill_values (the bug fix)
2. ✅ Multiple dtypes with tensor fill_values  
3. ✅ Python scalar fill_values like `42.0` (existing behavior)
4. ✅ Special float values like `float("-inf")` (transformer use case)

All tests are passing locally. Ready for another review when you have time!

Lucaskabela commented on torch/_dynamo/variables/torch.py:
```diff
@@ -834,11 +834,24 @@ def handle_addcdiv(self, tx: "InstructionTranslator", *args, **kwargs):
+                # Call fill_ method on the empty tensor
+                return empty_result.call_method(tx, "fill_", [fill_value], {})
+            else:
+                # For Python scalars and other non-tensor types, use default lowering
```
Ah, I thought this was supposed to default to the original logic:
```
                result = TorchInGraphFunctionVariable(
                    torch.ops.aten._local_scalar_dense
                ).call_function(tx, [fill_value], {})
                return TorchInGraphFunctionVariable(torch.full).call_function(
                    tx, [size, result], kwargs
```

Why did we change this to wrap_fx_proxy?

That is the source of the ci failure (arg `requires_grad` gets forwarded to the `aten.full.default` which doesn't exist

Lucaskabela reviewed (changes requested):
Curious why the CI signal didn't catch at land time - @atalman  any thoughts? Are these test that run less frequently?

amaldevh committed:
The previous fix caused infinite recursion by recursively calling
torch.full for scalar fill_values, which would hit the same handler
again indefinitely.

Fixed by only handling the TensorVariable case and returning None
for all other cases, which allows the code to fall through to the
default handling path that properly lowers to aten ops.

Lucaskabela reviewed (approved):
Thanks for quick fix - before landing, please run the command that previously failed and reply with the output.  Once you validate it is passing we can reland
